fe2-android/app/src/androidTest/java/com/github/dedis/popstellar/ui/detail/LaoDetailActivityTest.java

```diff
@@ -0,0 +1,182 @@
+      new ActivityScenarioRule<>(
+          IntentUtils.createIntent(
+              LaoDetailActivity.class,
+              new BundleBuilder()
+                  .putString(laoIdExtra(), LAO_ID)
+                  .putString(fragmentToOpenExtra(), laoDetailValue())
+                  .build()));
+
+  @Rule
+  public final RuleChain rule =
+      RuleChain.outerRule(MockitoJUnit.testRule(this))
+          .around(hiltAndroidRule)
+          .around(setupRule)
+          .around(activityScenarioRule);
+
+  @Test
+  public void homeButtonOpensHome() {
+    homeButton().perform(click());
+
+    assertEquals(
+        HomeActivity.class.getName(),
+        activityScenarioRule
+            .getScenario()
+            .getResult()
+            .getResultData()
+            .getComponent()
+            .getClassName());
+  }
+
+  @Test
+  public void identityButtonOpensIdentityTab() {
+    identityButton().perform(click());
+    fragmentContainer().check(matches(withChild(withId(identityFragmentId()))));
+  }
+
+  @Test
+  public void showPropertyButtonShowsConnectQRCode() {
+    showPropertiesButton().perform(click());
+
+    propertiesLayout().check(matches(isDisplayed()));
+
+    String expectedQRCode = gson.toJson(new ConnectToLao(requestFactory.getUrl(), LAO_ID));
+    connectQrCode().check(matches(withQrCode(expectedQRCode)));
+  }
+
+  @Test
+  public void witnessButtonShowsWitnessTab() {
+    witnessButton().perform(click());
+    fragmentContainer().check(matches(withChild(withId(witnessFragmentId()))));
+  }
+
+  // Matches an ImageView containing a QRCode with expected content
+  private Matcher<? super View> withQrCode(String expectedContent) {
+    return new TypeSafeMatcher<View>() {
+      @Override
+      protected boolean matchesSafely(View item) {
+        if (item instanceof ImageView) {
+          ImageView image = (ImageView) item;
+          Drawable drawable = image.getDrawable();
+
+          if (drawable instanceof BitmapDrawable) {
+            Bitmap qrcode = ((BitmapDrawable) drawable).getBitmap();
+            // Convert the QRCode to something zxing understands
+            int[] buffer = new int[qrcode.getWidth() * qrcode.getHeight()];
+            qrcode.getPixels(
+                buffer, 0, qrcode.getWidth(), 0, 0, qrcode.getWidth(), qrcode.getHeight());
+            LuminanceSource source =
+                new RGBLuminanceSource(qrcode.getWidth(), qrcode.getHeight(), buffer);
+            BinaryBitmap bitmap = new BinaryBitmap(new HybridBinarizer(source));
+            try {
+              // Parse the bitmap and check it against expected value
+              return expectedContent.equals(new QRCodeReader().decode(bitmap).getText());
+            } catch (NotFoundException | ChecksumException | FormatException e) {
+              throw new IllegalArgumentException("The provided image is not a valid QRCode", e);
+            }
+          }
+        }
+        return false;
```

Comment:
for my own understanding, what's the semantic difference between throwing and returning false, in a `TypeSafeMatcher` ?

Reply:
From my understanding, a boolean is returned if the data could be extracted. 
An exception is thrown when the data cannot be retrieved.

I made some changes to make it clearer.